### ISSUE TITLE
A few questions in the context of LLM

### ISSUE THREAD

**Issue by azsh1725** [NVIDIA/TensorRT#2864]
Hi! 

I have some questions. 

If I understand correctly, when creating an engine, TensorRT performs all the operations on the GPU. Therefore, if the hardware changes, a new engine needs to be generated. 

- Is it correct to assume that if the original model in the ONNX format does not fit into the GPU memory, it is not possible to convert it to TensorRT? My question pertains to large language models. I want to understand if it is possible to optimize them using TensorRT in any way.
- if the assumption from the previous question is correct, is it possible to create an engine using multiple identical GPUs to increase memory for usage by TensorRT.

Thank you in advance

**Comment by zerollzeng** [comment]
> If I understand correctly when creating an engine, TensorRT performs all the operations on the GPU. Therefore, if the hardware changes, a new engine needs to be generated.

Yes for now, I think we have some work on the engine forward compatibility, @oxana-nvidia should know more about it :)

> * Is it correct to assume that if the original model in the ONNX format does not fit into the GPU memory, it is not possible to convert it to TensorRT? My question pertains to large language models. I want to understand if it is possible to optimize them using TensorRT in any way.

I think it's case by case and we don't have a guarantee for it, no matter yes or not.

> * if the assumption from the previous question is correct, is it possible to create an engine using multiple identical GPUs to increase memory for usage by TensorRT.

Multi-device TRT is in our plan. IIRC it should come out in TRT 8.7. @oxana-nvidia please correct me if I'm wrong.

**Comment by oxana-nvidia** [comment]
Starting TRT 8.6, you can build a hardware compatible engine so you can build on one device and run it on another. Details and "how to start" points can be found in this GTC 2023 presentation https://www.nvidia.com/en-us/on-demand/session/gtcspring23-s51656/

Current requirement is to have all weights of the model on the GPU. We have some work in progress to relax this limitation. No information so far if and when we will release it publicly. Multi-device TRT is also work in progress, again no information if and when we will release it publicly. For LLM support @nvpohanh would be the right person to ask. Also you can read this blog https://developer.nvidia.com/blog/accelerated-inference-for-large-transformer-models-using-nvidia-fastertransformer-and-nvidia-triton-inference-server/

**Comment by nvpohanh** [comment]
> Is it correct to assume that if the original model in the ONNX format does not fit into the GPU memory, it is not possible to convert it to TensorRT? My question pertains to large language models. I want to understand if it is possible to optimize them using TensorRT in any way.

This is partially correct. You are correct that TRT currently only supports single-GPU inference, so if the ONNX model doesn't fit into single-GPU memory, it cannot run in TRT. (We are working on adding multi-GPU inference support, but that will take a while).

Meanwhile, it is still possible to run large language models using TRT if you manipulate the ONNX file. For example, if your GPT model has 96 decoder blocks, then you can break that into two ONNX files with 48 blocks each, and then you can run first 48 blocks on GPU0 and the last 48 blocks on GPU1. (this is what we called "pipeline parallelism") In this way, you can still use TRT to do multi-GPU inference.

**Comment by nvpohanh** [comment]
HuggingFace page has some docs about what pipeline-parallelism is: https://huggingface.co/docs/transformers/v4.15.0/parallelism#concepts

> My understanding was that creating an engine and inference are two separate and distinct stages. In particular, the creation of an engine involves analyzing the model graph and optimizing it under specific conditions, while inference simply entails launching a prepared engine.

Yes, this is 100% correct.

> Specifically, I was primarily interested in the process of creating an engine, rather than multi-GPU inference.

Yes, I was referring to engine creation. If your ONNX model is too large, TRT will not be able to create the engine. However, if you can break the ONNX model into several smaller ONNX files, TRT will be able to create one engine per file. And then at inference phase, you can simply connect them in series to perform the full inference.

**Comment by azsh1725** [comment]
@nvpohanh Ok I got it, thanks a lot for your help, i'll try.
